dynamicio/io/resource.py

```diff
@@ -0,0 +1,97 @@
+from abc import abstractmethod
+from copy import deepcopy
+from pathlib import Path
+from typing import Callable, List, Optional, Type
+
+import pandas as pd
+from pandera import SchemaModel
+from pydantic import BaseModel
+from uhura import Readable, Writable
+
+from dynamicio.inject import InjectionError, inject
+from dynamicio.io.serde import BaseSerde, PickleSerde
+
+
+def create_schema_validator(schema) -> Callable[[pd.DataFrame], pd.DataFrame]:
+    def validate_schema(df: pd.DataFrame):
+        return schema.validate(df)
+
+    return validate_schema
+
+
+class BaseResource(BaseModel, Readable[pd.DataFrame], Writable[pd.DataFrame]):  # Cacheable kaputt, ABC kaputt
+    """Base class for all resources.
+
+    :injectables: List of attributes that can be injected with format strings.
+    :test_path (optional): Path to the test data. If set, the resource will be read from and written to this path.
+    :pa_schema (optional): Pandera schema to validate the resource. If set, the resource will be validated before writing and after reading.
+    """
+
+    injectables: List[str]
+    test_path: Optional[Path] = None
+    pa_schema: Optional[Type[SchemaModel]] = None
+
+    def inject(self, **kwargs) -> "BaseResource":
+        # copy object
+        clone = deepcopy(self)
+        for injectable in self.injectables:
+            # inject attributes
+            value = getattr(clone, injectable)
+            if isinstance(value, str) or isinstance(value, Path) or value is None:
+                formatted_str = inject(value, **kwargs)
+                setattr(clone, injectable, formatted_str)
+
+            else:
+                raise InjectionError(f"Cannot inject {injectable} of type {type(value)}")
```

Comment:
nit: should we also add something like `in {self}` to make debugging in easier?

Reply:
Adding `{self.__class__.__name__}`